crates/wamr-sys/build.rs

```diff
@@ -14,12 +14,35 @@ fn main() {
     let wamr_root = wamr_root.join("wasm-micro-runtime");
     assert!(wamr_root.exists());
 
-    let is_espidf = env::var("CARGO_CFG_TARGET_OS").unwrap() != "espidf";
+    let is_espidf = env::var("CARGO_FEATURE_ESP_IDF").is_ok()
+        && env::var("CARGO_CFG_TARGET_OS").unwrap() == "espidf";
 
-    if is_espidf {
+    println!("cargo:rerun-if-env-changed=WAMR_BUILD_PLATFORM");
+    println!("cargo:rerun-if-env-changed=WAMR_SHARED_PLATFORM_CONFIG");
+    println!("cargo:rerun-if-env-changed=CARGO_FEATURE_ESP_IDF");
+
+    if is_espidf
+        && (env::var("WAMR_BUILD_PLATFORM").is_ok()
+            || env::var("WAMR_SHARED_PLATFORM_CONFIG").is_ok())
+    {
+        panic!("ESP-IDF build cannot use custom platform build (WAMR_BUILD_PLATFORM) or shared platform config (WAMR_SHARED_PLATFORM_CONFIG)");
```

Comment:
I am confused here. I though ESP-IDF building should use stuff in `product-mini/platforms/esp-idf/` in wamr repo. But the `panic!` seems suggest another way?

Reply:
IIUC, the new code suggests the build.rs should quit as soon as possible when realizing it is for an *esp-idf* build. 

If so, `is_espidf` means `cargo --target=espidf` and host platform is espidf.

Then we might want to check if the value `env::var("WAMR_BUILD_PLATFORM").unwrap() == "esp-idf"` instead of if `WAMR_BULD_PLATFORM` is defined only. I guess it could be something else. 

In order to avoid crossing compilation, might like
``` rust
if is_espidf || env::var("WAMR_BUILD_PLATFORM").unwrap() == "esp-idf"
```

`SHARED_PLATFORM_CONFIG` is assumed to match `WAMR_BUILD_PLATFORM`. So I guess we can ignore it. 

Please correct me, if I am wrong.

Reply:
The build script should exit immediately since we still need to generate the bindings themselves (which are the same for every platform), whic is performed at the end of the build script. However, the linking functions listed below are different and depend on the platform.

When building for ESP-IDF, the host platform can still be Linux, macOS, or Windows, but the target is `espidf` (e.g., `xtensa-esp32-espidf`, `riscv-espc3-espidf` etc.).

`WAMR_BUILD_PLATFORM` and `SHARED_PLATFORM_CONFIG` should not be defined when building for ESP-IDF since it doesn't make sense to declare them; everything is handled by `esp-idf-sys`.

For more clarity, here are the different possible build workflows:

| Target  | Feature | Env                                   | Compile?                          |
|---------|---------|---------------------------------------|-----------------------------------|
| esp-idf | esp-idf | None                                  | Yes                               |
| esp-idf | None    | `WAMR_BUILD_PLATFORM` / `SHARED_PLATFORM_CONFIG` | No: missing libc / esp-idf       |
| others  | None    | None                                  | Yes                               |
| others  | None    | `WAMR_BUILD_PLATFORM` / `SHARED_PLATFORM_CONFIG` | Yes                               |
| others  | esp-idf | None                                  | No: doesn’t make any sense        |

Reply:
IIUC, 
For scenario #2, the command is like `$cargo --target=espide build` ?
Fro scenario #4, the command is like `$ WAMR_BUILD_PLATFORM=others SHARED_PLATFORM_CONFIG=/path/to/others/shared_platform.cmake cargo --target=others` ?

Reply:
> The build script should exit immediately since we still need to generate 

should -> should not?

Reply:
IIUC, when the target is *esp-idf*, we can ignore `WAMR_BUILD_PLATFORM` and `SHARE_PPLATFORM_CONFIG`. Because L31~L64 is protected by `if !is_espidf` condition. So we can skip cmake compilation and goto bindings generation.